File: src/test/java/cube8540/oauth/authentication/credentials/security/UriSecurityMetadataSourceTest.java
package cube8540.oauth.authentication.credentials.security;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.security.access.ConfigAttribute;
import org.springframework.security.web.FilterInvocation;
import org.springframework.security.web.util.matcher.RequestMatcher;

import javax.servlet.http.HttpServletRequest;
import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import java.util.Set;

import static cube8540.oauth.authentication.credentials.security.SecurityApplicationTestHelper.makeAccessibleMetadata;
import static cube8540.oauth.authentication.credentials.security.SecurityApplicationTestHelper.makeAllMetadata;
import static cube8540.oauth.authentication.credentials.security.SecurityApplicationTestHelper.makeMetadata;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@DisplayName("URI 베이스 보안 메타 데이터 소스 테스트")
class UriSecurityMetadataSourceTest {

    @Test
    @DisplayName("클래스 인스터스화시 메타 데이터 초기화")
    void initializeMetadataWhenClassInstance() {
        Map<RequestMatcher, Collection<ConfigAttribute>> metadata = Collections.emptyMap();
        SecurityMetadataLoadService service = mock(SecurityMetadataLoadService.class);

        when(service.loadSecurityMetadata()).thenReturn(metadata);

        UriSecurityMetadataSource source = new UriSecurityMetadataSource(service);
        assertEquals(metadata, source.getMetadata());
    }

    @Test
    @DisplayName("모든 메타 데이터 반환")
    void returnAllMetadata() {
        HttpServletRequest request = mock(HttpServletRequest.class);
        Map<RequestMatcher, Collection<ConfigAttribute>> metadata = makeMetadata(request);
        SecurityMetadataLoadService service = mock(SecurityMetadataLoadService.class);

        when(service.loadSecurityMetadata()).thenReturn(metadata);

        UriSecurityMetadataSource source = new UriSecurityMetadataSource(service);
        Set<ConfigAttribute> excepted = makeAllMetadata();
        assertEquals(excepted, source.getAllConfigAttributes());
    }

    @Test
    @DisplayName("요청 받은 정보에 접근 가능한 메타 데이터 반환")
    void returnAccessibleMetadata() {
        FilterInvocation invocation = mock(FilterInvocation.class);
        HttpServletRequest request = mock(HttpServletRequest.class);
        Map<RequestMatcher, Collection<ConfigAttribute>> metadata = makeMetadata(request);
        SecurityMetadataLoadService service = mock(SecurityMetadataLoadService.class);

        when(invocation.getRequest()).thenReturn(request);
        when(service.loadSecurityMetadata()).thenReturn(metadata);

        UriSecurityMetadataSource source = new UriSecurityMetadataSource(service);
        Set<ConfigAttribute> excepted = makeAccessibleMetadata();
        assertEquals(excepted, source.getAttributes(invocation));
    }

    @Test
    @DisplayName("메타 데이터 리로드")
    void reloadMetadata() {
        Map<RequestMatcher, Collection<ConfigAttribute>> originalMetadata = Collections.emptyMap();
        Map<RequestMatcher, Collection<ConfigAttribute>> reloadMetadata = Collections.singletonMap(mock(RequestMatcher.class), Collections.emptyList());
        SecurityMetadataLoadService metadataLoadService = mock(SecurityMetadataLoadService.class);
        when(metadataLoadService.loadSecurityMetadata()).thenReturn(originalMetadata);
        UriSecurityMetadataSource source = new UriSecurityMetadataSource(metadataLoadService);

        when(metadataLoadService.loadSecurityMetadata()).thenReturn(reloadMetadata);

        source.reload();
        assertEquals(reloadMetadata, source.getMetadata());
    }

    @Test
    @DisplayName("지원 되는 클래스 확인")
    void checkSupportedClass() {
        SecurityMetadataLoadService service = mock(SecurityMetadataLoadService.class);

        UriSecurityMetadataSource source = new UriSecurityMetadataSource(service);

        assertTrue(source.supports(FilterInvocation.class));
    }

}

File: src/test/kotlin/cube8540/oauth/authentication/security/UriSecurityMetadataSourceTest.kt
package cube8540.oauth.authentication.security

import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.security.access.ConfigAttribute
import org.springframework.security.access.SecurityConfig
import org.springframework.security.web.FilterInvocation
import org.springframework.security.web.util.matcher.AntPathRequestMatcher
import org.springframework.security.web.util.matcher.RequestMatcher
import javax.servlet.http.HttpServletRequest

class UriSecurityMetadataSourceTest {
    private val pool = ('a'..'z').map { it.toString() }

    private val metadataLoadService: SecurityMetadataLoadService = mockk(relaxed = true)

    @Nested
    inner class InitializationTest {
        private val metadata: Map<RequestMatcher, Collection<ConfigAttribute>> = hashMapOf(
            mockk<RequestMatcher>(relaxed = true) to listOf(SecurityConfig(pool.random()), SecurityConfig(pool.random())),
            mockk<RequestMatcher>(relaxed = true) to listOf(SecurityConfig(pool.random()), SecurityConfig(pool.random())),
            mockk<RequestMatcher>(relaxed = true) to listOf(SecurityConfig(pool.random()), SecurityConfig(pool.random()))
        )

        init {
            every { metadataLoadService.loadSecurityMetadata() } returns metadata
        }

        @Test
        fun `initialize metadata`() {
            val metadataSource = UriSecurityMetadataSource(metadataLoadService)

            val result = metadataSource.metadata
            assertThat(result).isEqualTo(metadata)
        }
    }

    @Nested
    inner class GetAllAttributeTest {
        private val requestMatcherA = AntPathRequestMatcher(pool.random())
        private val metadataA = listOf(SecurityConfig(pool.random()), SecurityConfig(pool.random()), SecurityConfig(pool.random()))

        private val requestMatcherB = AntPathRequestMatcher(pool.random())
        private val metadataB = listOf(SecurityConfig(pool.random()), SecurityConfig(pool.random()), SecurityConfig(pool.random()))

        private val requestMatcherC = AntPathRequestMatcher(pool.random())
        private val metadataC = listOf(SecurityConfig(pool.random()), SecurityConfig(pool.random()), SecurityConfig(pool.random()))

        private val metadata: Map<RequestMatcher, Collection<ConfigAttribute>> = hashMapOf(
            requestMatcherA to metadataA,
            requestMatcherB to metadataB,
            requestMatcherC to metadataC
        )

        init {
            every { metadataLoadService.loadSecurityMetadata() } returns metadata
        }

        @Test
        fun `return all metadata`() {
            val metadataSource = UriSecurityMetadataSource(metadataLoadService)

            val result = metadataSource.allConfigAttributes
            val expected = (metadataA + metadataB + metadataC).toSet()

            assertThat(result).isEqualTo(expected)
        }
    }

    @Nested
    inner class GetAccessibleAttributeTest {
        private val requestMatcherA: RequestMatcher = mockk()
        private val metadataAttributeA = listOf(SecurityConfig(pool.random()), SecurityConfig(pool.random()), SecurityConfig(pool.random()))

        private val requestMatcherB: RequestMatcher = mockk()
        private val metadataAttributeB = listOf(SecurityConfig(pool.random()), SecurityConfig(pool.random()), SecurityConfig(pool.random()))

        private val requestMatcherC: RequestMatcher = mockk()
        private val metadataAttributeC = listOf(SecurityConfig(pool.random()), SecurityConfig(pool.random()), SecurityConfig(pool.random()))

        init {
            every { metadataLoadService.loadSecurityMetadata() } returns hashMapOf(
                requestMatcherA to metadataAttributeA,
                requestMatcherB to metadataAttributeB,
                requestMatcherC to metadataAttributeC
            )
        }

        @Test
        fun `accessible metadata not found`() {
            val servletRequest: HttpServletRequest = mockk()
            val invocation: FilterInvocation = mockk {
                every { request } returns servletRequest
            }
            val metadataSource = UriSecurityMetadataSource(metadataLoadService)

            every { requestMatcherA.matches(servletRequest) } returns false
            every { requestMatcherB.matches(servletRequest) } returns false
            every { requestMatcherC.matches(servletRequest) } returns false

            val result = metadataSource.getAttributes(invocation)
            assertThat(result).isEmpty()
        }

        @Test
        fun `get accessible metadata`() {
            val servletRequest: HttpServletRequest = mockk()
            val invocation: FilterInvocation = mockk {
                every { request } returns servletRequest
            }
            val metadataSource = UriSecurityMetadataSource(metadataLoadService)

            every { requestMatcherA.matches(servletRequest) } returns true
            every { requestMatcherB.matches(servletRequest) } returns true
            every { requestMatcherC.matches(servletRequest) } returns true

            val result = metadataSource.getAttributes(invocation)
            val excepted = (metadataAttributeA + metadataAttributeB + metadataAttributeC).toSet()
            assertThat(result).isEqualTo(excepted)
        }
    }

    @Nested
    inner class ReloadTest {
        private val metadata: Map<RequestMatcher, Collection<ConfigAttribute>> = hashMapOf(
            mockk<RequestMatcher>(relaxed = true) to listOf(SecurityConfig(pool.random()), SecurityConfig(pool.random()))
        )
        private val reloadMetadata: Map<RequestMatcher, Collection<ConfigAttribute>> = hashMapOf(
            mockk<RequestMatcher>(relaxed = true) to listOf(SecurityConfig(pool.random()), SecurityConfig(pool.random()))
        )

        init {
            every { metadataLoadService.loadSecurityMetadata() } returns metadata
        }

        @Test
        fun `reload metadata`() {
            val metadataSource = UriSecurityMetadataSource(metadataLoadService)

            every { metadataLoadService.loadSecurityMetadata() } returns reloadMetadata

            metadataSource.reload()
            assertThat(metadataSource.metadata).isEqualTo(reloadMetadata)
        }
    }

    @Nested
    inner class CheckSupportTest {

        @Test
        fun `check supported class`() {
            val metadataSource = UriSecurityMetadataSource(metadataLoadService)

            val result = metadataSource.supports(FilterInvocation::class.java)
            assertThat(result).isTrue
        }
    }
}